### PR TITLE
Migrate remaining old decl_* macros to the new pallet attribute macros

### PR DESCRIPTION
**Pull request by koushiro**
## What does it do?

Migrate remaining old `decl_`* macros to the new `#[pallet]` attribute macros

part of #12248

---

polkadot address: 15XuanNimo5951s3RjFTPX1AvYVnCwfr3SDhb3AV4fQF3LpK

### CODE REVIEW

File: frame/benchmarking/src/tests.rs [hunk 1]
```diff
@@ -36,7 +36,7 @@ mod pallet_test {
 
 	#[pallet::pallet]
 	#[pallet::generate_store(pub(super) trait Store)]
-	pub struct Pallet<T>(_);
+	pub struct Pallet<T>(PhantomData<T>);
```
Comment (bkchr) [inline frame/benchmarking/src/tests.rs:39]:
```suggestion
	pub struct Pallet<T>(_);
```

File: test-utils/runtime/src/system.rs [hunk 2]
```diff
@@ -39,19 +38,94 @@ use sp_std::prelude::*;
...
+		}
+		const STORAGE_PREFIX: &'static str = "Authorities";
+	}
+	pub type Authorities = StorageValue<AuthoritiesPrefix, Vec<AuthorityId>, ValueQuery>;
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig<T: Config> {
+		pub authorities: Vec<AuthorityId>,
+		pub marker: PhantomData<T>,
+	}
```
Comment (bkchr) [inline test-utils/runtime/src/system.rs:115]:
```suggestion
	pub struct GenesisConfig {
		pub authorities: Vec<AuthorityId>,
	}
```

Just don't make it generic, if you don't need it.

File: utils/frame/rpc/support/src/lib.rs [hunk 3]
```diff
@@ -34,33 +34,73 @@ use sp_storage::{StorageData, StorageKey};
...
+/// 	pub type Actions = StorageMap<ActionsPrefix, Blake2_128Concat, u64, Loc>;
+///
+/// 	pub struct PrefabPrefix;
+/// 	impl StorageInstance for PrefabPrefix {
+/// 		fn pallet_prefix() -> &'static str {
+/// 			"TestRuntime"
+/// 		}
+/// 		const STORAGE_PREFIX: &'static str = "Prefab";
+/// 	}
+/// 	pub type Prefab = StorageDoubleMap<
```
Comment (bkchr) [inline utils/frame/rpc/support/src/lib.rs:99]: Why don't you use `pallet::storage` here and for the other types?

Comment (koushiro) [reply utils/frame/rpc/support/src/lib.rs:99]: If I use `pallet::storage` here, I will need to use `construct_runtime!` to generate some Runtime trivials,
and add `impl frame_system::Config for Runtime` and `impl Config for Runtime`.

And you must add generic for the storage types if you use `pallet::storage`, like:

```
error: Invalid type def generics: expected `T` or `T: Config` or `T, I = ()` or `T: Config<I>, I: 'static = ()`
  --> src/lib.rs:73:11
   |
42 |     pub type LastActionId = StorageValue<_, u64, ValueQuery>;
   |              ^^^^^^^^^^^^
```

File: frame/benchmarking/src/tests.rs (hunk 1 above)
Comment (koushiro) [reply frame/benchmarking/src/tests.rs:39]: Why not use `struct Pallet<T>(PhantomData<T>);`?

`struct Pallet<T>(_);` is not standard rust syntax

Comment (bkchr) [reply frame/benchmarking/src/tests.rs:39]: Yeah I know. This syntax was added as some shortcut.

Comment (shawntabrizi) [reply frame/benchmarking/src/tests.rs:39]: i would not be against removing these shortcuts if you find it confusing. End of the day, no one writes this line by hand, they just copy some template.

File: test-utils/runtime/src/system.rs [hunk 4]
```diff
@@ -39,19 +38,87 @@ use sp_std::prelude::*;
...
+	#[pallet::config]
+	pub trait Config: frame_system::Config {}
+
+	pub struct ExtrinsicDataPrefix;
+	impl StorageInstance for ExtrinsicDataPrefix {
+		fn pallet_prefix() -> &'static str {
+			"TestRuntime"
+		}
+		const STORAGE_PREFIX: &'static str = "ExtrinsicData";
+	}
```
Comment (bkchr) [inline test-utils/runtime/src/system.rs:61]: Again, I don't get why we need this?

Comment (koushiro) [reply test-utils/runtime/src/system.rs]: I'm not sure if anyone else is using this public module directly, so I've kept the changes to a minimum, if you mind I'll use ·pallet::storage· instead and add generic to the storage types.

**Review by bkchr — APPROVED** [review]
CI isn't happy, but looks good otherwise!

Ty

**Comment by koushiro** [comment]
@KiChjang @shawntabrizi PTAL

**Comment by bkchr** [comment]
@koushiro tests are still failing.

**Comment by koushiro** [comment]
> @koushiro tests are still failing.

Fixed

File: client/service/test/src/client/mod.rs [hunk 5]
```diff
@@ -1670,6 +1670,7 @@ fn storage_keys_iter_works() {
 			"1a560ecfd2a62c2b8521ef149d0804eb621050e3988ed97dca55f0d7c3e6aa34",
 			"1d66850d32002979d67dd29dc583af5b2ae2a1f71c1f35ad90fff122be7a3824",
 			"237498b98d8803334286e9f0483ef513098dd3c1c22ca21c4dc155b4ef6cc204",
+			"26aa394eea5630e07c48ae0c9558cef75e0621c4869aa60c02be9adcc98a0d1d",
```
Comment (kianenigma) [inline client/service/test/src/client/mod.rs:1673]: strange changes. Why?

**Review by kianenigma — APPROVED** [review]

**Comment by kianenigma** [comment]
By the end of this PR, we should have no more use of `decl_*` in our codebase, so we should mark them as deprecated? @KiChjang WDYT?

File: client/service/test/src/client/mod.rs (hunk 5 above)
Comment (koushiro) [reply client/service/test/src/client/mod.rs:1673]: Before this changes, the system module of system_test_runtime use `TestRuntime` as pallet prefix, but now it use `System` instead.

**Comment by bkchr** [comment]
bot rebase

**Comment by bkchr** [comment]
bot merge

**Comment by bkchr** [comment]
bot merge

**Comment by KiChjang** [comment]
> By the end of this PR, we should have no more use of `decl_*` in our codebase, so we should mark them as deprecated? @KiChjang WDYT?

We already have a PR for that: #12326, although I must say that perhaps what we should really do is remove them rather than simply deprecating them.

**Pull request merged**